In [1]:
import pandas as pd

cards_df = pd.read_csv("cards_data.csv")
cards_df.head()

,id,client_id,card_brand,card_type,card_number,expires,cvv,has_chip,num_cards_issued,credit_limit,acct_open_date,year_pin_last_changed,card_on_dark_web
0,4524,825,Visa,Debit,4344676511950444,12/2022,623,YES,2,$24295,09/2002,2008,No
1,2731,825,Visa,Debit,4956965974959986,12/2020,393,YES,2,$21968,04/2014,2014,No
2,3701,825,Visa,Debit,4582313478255491,02/2024,719,YES,2,$46414,07/2003,2004,No
3,42,825,Visa,Credit,4879494103069057,08/2024,693,NO,1,$12400,01/2003,2012,No
4,4659,825,Mastercard,Debit (Prepaid),5722874738736011,03/2009,75,YES,1,$28,09/2008,2009,No


- **card_brand** → The type of network or card it belongs to 
- **card_type** → Card type 
- **expires** → Card expiration date 
- **cvv** → The PIN found on each card  
- **has_chip** → Does the card have a smart chip? 
- **credit_limit** → Card credit limit
- **acct_open_date** → Date the account was opened or the card was first issued 
- **year_pin_last_changed** → Last year the PIN was changed 




In [2]:
user_df = pd.read_csv("users_data.csv")
user_df

,id,current_age,retirement_age,birth_year,birth_month,gender,address,latitude,longitude,per_capita_income,yearly_income,total_debt,credit_score,num_credit_cards
0,825,53,66,1966,11,Female,462 Rose Lane,34.15,-117.76,$29278,$59696,$127613,787,5
1,1746,53,68,1966,12,Female,3606 Federal Boulevard,40.76,-73.74,$37891,$77254,$191349,701,5
2,1718,81,67,1938,11,Female,766 Third Drive,34.02,-117.89,$22681,$33483,$196,698,5
3,708,63,63,1957,1,Female,3 Madison Street,40.71,-73.99,$163145,$249925,$202328,722,4
4,1164,43,70,1976,9,Male,9620 Valley Stream Drive,37.76,-122.44,$53797,$109687,$183855,675,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1995,986,32,70,1987,7,Male,6577 Lexington Lane,40.65,-73.58,$23550,$48010,$87837,703,3
1996,1944,62,65,1957,11,Female,2 Elm Drive,38.95,-84.54,$24218,$49378,$104480,740,4
1997,185,47,67,1973,1,Female,276 Fifth Boulevard,40.66,-74.19,$15175,$30942,$71066,779,3
1998,1007,66,60,1954,2,Male,259 Valley Boulevard,40.24,-76.92,$25336,$54654,$27241,618,1


- **retirement_age** → عمر تقاعد الشخص عن العمل  
- **latitude** → خط العرض بيحدد المكان بالنسبة للشمال او الجنوب عن خط الاستواء  
- **longitude** → خط الطول بيحدد المكان بالنسبة للشرق او الغرب عن خط غرينتش  
- **per_capita_income** → نصيب الفرد من الدخل (متوسط دخل الفرد)  
- **total_debt** → المبلغ الكلي اللي الشخص مديون بيه 

In [3]:
user_df.head()

,id,current_age,retirement_age,birth_year,birth_month,gender,address,latitude,longitude,per_capita_income,yearly_income,total_debt,credit_score,num_credit_cards
0,825,53,66,1966,11,Female,462 Rose Lane,34.15,-117.76,$29278,$59696,$127613,787,5
1,1746,53,68,1966,12,Female,3606 Federal Boulevard,40.76,-73.74,$37891,$77254,$191349,701,5
2,1718,81,67,1938,11,Female,766 Third Drive,34.02,-117.89,$22681,$33483,$196,698,5
3,708,63,63,1957,1,Female,3 Madison Street,40.71,-73.99,$163145,$249925,$202328,722,4
4,1164,43,70,1976,9,Male,9620 Valley Stream Drive,37.76,-122.44,$53797,$109687,$183855,675,1


In [4]:
user_df.tail()

,id,current_age,retirement_age,birth_year,birth_month,gender,address,latitude,longitude,per_capita_income,yearly_income,total_debt,credit_score,num_credit_cards
1995,986,32,70,1987,7,Male,6577 Lexington Lane,40.65,-73.58,$23550,$48010,$87837,703,3
1996,1944,62,65,1957,11,Female,2 Elm Drive,38.95,-84.54,$24218,$49378,$104480,740,4
1997,185,47,67,1973,1,Female,276 Fifth Boulevard,40.66,-74.19,$15175,$30942,$71066,779,3
1998,1007,66,60,1954,2,Male,259 Valley Boulevard,40.24,-76.92,$25336,$54654,$27241,618,1
1999,1110,21,60,1998,11,Female,472 Ocean View Street,42.86,-71.48,$32325,$65909,$181261,673,2


In [5]:
user_df.shape

(2000, 14)

In [6]:
user_df.columns

Index(['id', 'current_age', 'retirement_age', 'birth_year', 'birth_month',
       'gender', 'address', 'latitude', 'longitude', 'per_capita_income',
       'yearly_income', 'total_debt', 'credit_score', 'num_credit_cards'],
      dtype='object')

In [7]:
user_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 2000 non-null   int64  
 1   current_age        2000 non-null   int64  
 2   retirement_age     2000 non-null   int64  
 3   birth_year         2000 non-null   int64  
 4   birth_month        2000 non-null   int64  
 5   gender             2000 non-null   object 
 6   address            2000 non-null   object 
 7   latitude           2000 non-null   float64
 8   longitude          2000 non-null   float64
 9   per_capita_income  2000 non-null   object 
 10  yearly_income      2000 non-null   object 
 11  total_debt         2000 non-null   object 
 12  credit_score       2000 non-null   int64  
 13  num_credit_cards   2000 non-null   int64  
dtypes: float64(2), int64(7), object(5)
memory usage: 218.9+ KB


In [8]:
user_df.describe()

,id,current_age,retirement_age,birth_year,birth_month,latitude,longitude,credit_score,num_credit_cards
count,2000.000000,2000.000000,2000.000000,2000.000000,2000.000000,2000.000000,2000.000000,2000.000000,2000.000000
mean,999.500000,45.391500,66.237500,1973.803000,6.439000,37.389225,-91.554765,709.734500,3.073000
std,577.494589,18.414092,3.628867,18.421234,3.565338,5.114324,16.283293,67.221949,1.637379
min,0.000000,18.000000,50.000000,1918.000000,1.000000,20.880000,-159.410000,480.000000,1.000000
25%,499.750000,30.000000,65.000000,1961.000000,3.000000,33.837500,-97.395000,681.000000,2.000000
50%,999.500000,44.000000,66.000000,1975.000000,7.000000,38.250000,-86.440000,711.500000,3.000000
75%,1499.250000,58.000000,68.000000,1989.000000,10.000000,41.200000,-80.130000,753.000000,4.000000
max,1999.000000,101.000000,79.000000,2002.000000,12.000000,61.200000,-68.670000,850.000000,9.000000


In [9]:
# outlier max by current_age
Q3 = 58.000000
IQR = 58.000000 - 30.000000
maxx = Q3 + 1.5 * IQR
maxx

100.0

In [10]:
outlier_max = user_df[user_df["current_age"] > maxx]
outlier_max   # ----------> there is one person whose age greater than 100

,id,current_age,retirement_age,birth_year,birth_month,gender,address,latitude,longitude,per_capita_income,yearly_income,total_debt,credit_score,num_credit_cards
300,1424,101,68,1918,5,Female,887 Third Boulevard,35.13,-112.99,$13086,$15348,$1396,761,4


In [11]:
user_df['current_age'].sort_values(ascending = False) # ----> here the first value "outlier"

300     101
1961     99
1101     98
983      98
1753     94
       ... 
1942     18
379      18
1165     18
1138     18
1626     18
Name: current_age, Length: 2000, dtype: int64

In [12]:
user_df.duplicated().sum()

0

In [13]:
user_df.isna().sum()

id                   0
current_age          0
retirement_age       0
birth_year           0
birth_month          0
gender               0
address              0
latitude             0
longitude            0
per_capita_income    0
yearly_income        0
total_debt           0
credit_score         0
num_credit_cards     0
dtype: int64

In [14]:
user_df[['current_age','gender']]

,current_age,gender
0,53,Female
1,53,Female
2,81,Female
3,63,Female
4,43,Male
...,...,...
1995,32,Male
1996,62,Female
1997,47,Female
1998,66,Male


In [15]:
user_df.head()

,id,current_age,retirement_age,birth_year,birth_month,gender,address,latitude,longitude,per_capita_income,yearly_income,total_debt,credit_score,num_credit_cards
0,825,53,66,1966,11,Female,462 Rose Lane,34.15,-117.76,$29278,$59696,$127613,787,5
1,1746,53,68,1966,12,Female,3606 Federal Boulevard,40.76,-73.74,$37891,$77254,$191349,701,5
2,1718,81,67,1938,11,Female,766 Third Drive,34.02,-117.89,$22681,$33483,$196,698,5
3,708,63,63,1957,1,Female,3 Madison Street,40.71,-73.99,$163145,$249925,$202328,722,4
4,1164,43,70,1976,9,Male,9620 Valley Stream Drive,37.76,-122.44,$53797,$109687,$183855,675,1


In [16]:
user_df['gender'].value_counts()

gender
Female    1016
Male       984
Name: count, dtype: int64

In [17]:
# remove $ to convert dtype from object to numeric
user_df['per_capita_income'] = user_df['per_capita_income'].replace(['\$'],'', regex = True).astype(float)
user_df['yearly_income'] = user_df['yearly_income'].replace(['\$'],'', regex = True).astype(float)
user_df['total_debt'] = user_df['total_debt'].replace(['\$'],'', regex = True).astype(float)

In [18]:
user_df.head()

,id,current_age,retirement_age,birth_year,birth_month,gender,address,latitude,longitude,per_capita_income,yearly_income,total_debt,credit_score,num_credit_cards
0,825,53,66,1966,11,Female,462 Rose Lane,34.15,-117.76,29278.0,59696.0,127613.0,787,5
1,1746,53,68,1966,12,Female,3606 Federal Boulevard,40.76,-73.74,37891.0,77254.0,191349.0,701,5
2,1718,81,67,1938,11,Female,766 Third Drive,34.02,-117.89,22681.0,33483.0,196.0,698,5
3,708,63,63,1957,1,Female,3 Madison Street,40.71,-73.99,163145.0,249925.0,202328.0,722,4
4,1164,43,70,1976,9,Male,9620 Valley Stream Drive,37.76,-122.44,53797.0,109687.0,183855.0,675,1


In [19]:
user_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 2000 non-null   int64  
 1   current_age        2000 non-null   int64  
 2   retirement_age     2000 non-null   int64  
 3   birth_year         2000 non-null   int64  
 4   birth_month        2000 non-null   int64  
 5   gender             2000 non-null   object 
 6   address            2000 non-null   object 
 7   latitude           2000 non-null   float64
 8   longitude          2000 non-null   float64
 9   per_capita_income  2000 non-null   float64
 10  yearly_income      2000 non-null   float64
 11  total_debt         2000 non-null   float64
 12  credit_score       2000 non-null   int64  
 13  num_credit_cards   2000 non-null   int64  
dtypes: float64(5), int64(7), object(2)
memory usage: 218.9+ KB


In [20]:
user_df['yearly_income'].sum()

91431764.0

In [21]:
user_df['total_debt'].sum()

127419388.0

In [22]:
user_df['per_capita_income'].sum()

46283856.0

In [23]:
user_df.groupby('gender')['yearly_income'].mean()

gender
Female    46048.314961
Male      45372.638211
Name: yearly_income, dtype: float64

In [24]:
user_df.groupby('gender')['per_capita_income'].sum()

gender
Female    23771580.0
Male      22512276.0
Name: per_capita_income, dtype: float64

In [25]:
user_df.head()

,id,current_age,retirement_age,birth_year,birth_month,gender,address,latitude,longitude,per_capita_income,yearly_income,total_debt,credit_score,num_credit_cards
0,825,53,66,1966,11,Female,462 Rose Lane,34.15,-117.76,29278.0,59696.0,127613.0,787,5
1,1746,53,68,1966,12,Female,3606 Federal Boulevard,40.76,-73.74,37891.0,77254.0,191349.0,701,5
2,1718,81,67,1938,11,Female,766 Third Drive,34.02,-117.89,22681.0,33483.0,196.0,698,5
3,708,63,63,1957,1,Female,3 Madison Street,40.71,-73.99,163145.0,249925.0,202328.0,722,4
4,1164,43,70,1976,9,Male,9620 Valley Stream Drive,37.76,-122.44,53797.0,109687.0,183855.0,675,1


In [26]:
# Debt_income_ratio column clarifies if a person in debt beyond his means or not
user_df['Debt_income_ratio'] = user_df['total_debt'] / user_df['per_capita_income']
user_df = user_df.round(2)
user_df

,id,current_age,retirement_age,birth_year,birth_month,gender,address,latitude,longitude,per_capita_income,yearly_income,total_debt,credit_score,num_credit_cards,Debt_income_ratio
0,825,53,66,1966,11,Female,462 Rose Lane,34.15,-117.76,29278.0,59696.0,127613.0,787,5,4.36
1,1746,53,68,1966,12,Female,3606 Federal Boulevard,40.76,-73.74,37891.0,77254.0,191349.0,701,5,5.05
2,1718,81,67,1938,11,Female,766 Third Drive,34.02,-117.89,22681.0,33483.0,196.0,698,5,0.01
3,708,63,63,1957,1,Female,3 Madison Street,40.71,-73.99,163145.0,249925.0,202328.0,722,4,1.24
4,1164,43,70,1976,9,Male,9620 Valley Stream Drive,37.76,-122.44,53797.0,109687.0,183855.0,675,1,3.42
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1995,986,32,70,1987,7,Male,6577 Lexington Lane,40.65,-73.58,23550.0,48010.0,87837.0,703,3,3.73
1996,1944,62,65,1957,11,Female,2 Elm Drive,38.95,-84.54,24218.0,49378.0,104480.0,740,4,4.31
1997,185,47,67,1973,1,Female,276 Fifth Boulevard,40.66,-74.19,15175.0,30942.0,71066.0,779,3,4.68
1998,1007,66,60,1954,2,Male,259 Valley Boulevard,40.24,-76.92,25336.0,54654.0,27241.0,618,1,1.08


In [27]:
user_df[user_df['Debt_income_ratio'] > 1]

,id,current_age,retirement_age,birth_year,birth_month,gender,address,latitude,longitude,per_capita_income,yearly_income,total_debt,credit_score,num_credit_cards,Debt_income_ratio
0,825,53,66,1966,11,Female,462 Rose Lane,34.15,-117.76,29278.0,59696.0,127613.0,787,5,4.36
1,1746,53,68,1966,12,Female,3606 Federal Boulevard,40.76,-73.74,37891.0,77254.0,191349.0,701,5,5.05
3,708,63,63,1957,1,Female,3 Madison Street,40.71,-73.99,163145.0,249925.0,202328.0,722,4,1.24
4,1164,43,70,1976,9,Male,9620 Valley Stream Drive,37.76,-122.44,53797.0,109687.0,183855.0,675,1,3.42
6,1075,36,67,1983,12,Female,5695 Fifth Street,38.22,-85.74,25258.0,51500.0,102286.0,672,3,4.05
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1995,986,32,70,1987,7,Male,6577 Lexington Lane,40.65,-73.58,23550.0,48010.0,87837.0,703,3,3.73
1996,1944,62,65,1957,11,Female,2 Elm Drive,38.95,-84.54,24218.0,49378.0,104480.0,740,4,4.31
1997,185,47,67,1973,1,Female,276 Fifth Boulevard,40.66,-74.19,15175.0,30942.0,71066.0,779,3,4.68
1998,1007,66,60,1954,2,Male,259 Valley Boulevard,40.24,-76.92,25336.0,54654.0,27241.0,618,1,1.08


In [28]:
user_df['high_risk'] = user_df['Debt_income_ratio'] > 1

In [29]:
user_df

,id,current_age,retirement_age,birth_year,birth_month,gender,address,latitude,longitude,per_capita_income,yearly_income,total_debt,credit_score,num_credit_cards,Debt_income_ratio,high_risk
0,825,53,66,1966,11,Female,462 Rose Lane,34.15,-117.76,29278.0,59696.0,127613.0,787,5,4.36,True
1,1746,53,68,1966,12,Female,3606 Federal Boulevard,40.76,-73.74,37891.0,77254.0,191349.0,701,5,5.05,True
2,1718,81,67,1938,11,Female,766 Third Drive,34.02,-117.89,22681.0,33483.0,196.0,698,5,0.01,False
3,708,63,63,1957,1,Female,3 Madison Street,40.71,-73.99,163145.0,249925.0,202328.0,722,4,1.24,True
4,1164,43,70,1976,9,Male,9620 Valley Stream Drive,37.76,-122.44,53797.0,109687.0,183855.0,675,1,3.42,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1995,986,32,70,1987,7,Male,6577 Lexington Lane,40.65,-73.58,23550.0,48010.0,87837.0,703,3,3.73,True
1996,1944,62,65,1957,11,Female,2 Elm Drive,38.95,-84.54,24218.0,49378.0,104480.0,740,4,4.31,True
1997,185,47,67,1973,1,Female,276 Fifth Boulevard,40.66,-74.19,15175.0,30942.0,71066.0,779,3,4.68,True
1998,1007,66,60,1954,2,Male,259 Valley Boulevard,40.24,-76.92,25336.0,54654.0,27241.0,618,1,1.08,True


In [30]:
user_df[['credit_score','num_credit_cards']].sort_values(by = 'credit_score', ascending=False)

,credit_score,num_credit_cards
490,850,6
1729,850,1
1313,850,6
773,850,4
1527,850,5
...,...,...
806,491,3
1510,490,2
908,489,3
150,488,2


In [31]:
user_df[user_df['credit_score'] < 600]

,id,current_age,retirement_age,birth_year,birth_month,gender,address,latitude,longitude,per_capita_income,yearly_income,total_debt,credit_score,num_credit_cards,Debt_income_ratio,high_risk
14,1590,48,62,1971,5,Female,781 East Street,29.45,-98.50,10059.0,20513.0,32509.0,599,1,3.23,True
46,1134,18,59,2002,2,Female,7 11th Drive,29.45,-98.50,12677.0,25847.0,46379.0,568,1,3.66,True
67,898,90,66,1929,6,Female,145 River Drive,40.59,-75.47,12427.0,19893.0,1712.0,566,3,0.14,False
132,1055,37,69,1982,4,Male,3472 11th Avenue,28.50,-81.37,15137.0,30865.0,55079.0,594,1,3.64,True
150,1987,63,62,1956,9,Male,786 12th Drive,42.13,-87.92,23098.0,33686.0,24997.0,488,2,1.08,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1888,1168,51,68,1968,10,Male,207 Ocean View Street,40.67,-74.42,53790.0,109673.0,242379.0,505,1,4.51,True
1905,897,30,64,1989,8,Female,847 Martin Luther King Lane,31.07,-83.19,13238.0,26996.0,33332.0,501,2,2.52,True
1913,1535,59,64,1960,8,Male,8 Pine Street,38.87,-82.99,16336.0,33311.0,65822.0,590,2,4.03,True
1951,1347,64,64,1955,5,Female,893 Littlewood Lane,35.73,-77.92,23756.0,55418.0,20950.0,590,3,0.88,False


In [32]:
user_df.groupby('num_credit_cards')['credit_score'].mean()

num_credit_cards
1    682.778846
2    700.912371
3    717.247216
4    718.484043
5    726.140777
6    733.580952
7    746.375000
8    749.000000
9    695.333333
Name: credit_score, dtype: float64

In [33]:
user_df['num_credit_cards'].value_counts()

num_credit_cards
3    449
1    416
2    388
4    376
5    206
6    105
7     40
8     17
9      3
Name: count, dtype: int64

In [34]:
user_df['credit_score'].value_counts()

credit_score
850    34
689    24
685    23
703    22
684    22
       ..
488     1
845     1
834     1
617     1
501     1
Name: count, Length: 321, dtype: int64

In [35]:
user_df[user_df['retirement_age'] < 60]

,id,current_age,retirement_age,birth_year,birth_month,gender,address,latitude,longitude,per_capita_income,yearly_income,total_debt,credit_score,num_credit_cards,Debt_income_ratio,high_risk
20,309,28,57,1991,8,Male,2473 Lake Avenue,32.07,-82.91,12321.0,25122.0,43205.0,819,3,3.51,True
27,775,37,58,1982,9,Male,776 Norfolk Boulevard,34.17,-118.46,21402.0,43638.0,104052.0,627,1,4.86,True
35,34,41,55,1978,8,Male,7467 Spruce Drive,38.48,-121.34,25431.0,51854.0,72162.0,737,2,2.84,True
43,1231,26,52,1994,1,Female,613 Little Creek Lane,26.41,-81.42,10819.0,22066.0,38967.0,842,1,3.60,True
46,1134,18,59,2002,2,Female,7 11th Drive,29.45,-98.50,12677.0,25847.0,46379.0,568,1,3.66,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1820,1794,49,58,1970,12,Male,469 Catherine Avenue,29.45,-98.50,36852.0,75137.0,175263.0,631,1,4.76,True
1868,795,40,58,1979,3,Female,73 Oak Street,32.79,-96.76,37234.0,75918.0,94579.0,759,2,2.54,True
1907,1781,29,53,1990,3,Male,73 South Lane,42.93,-75.46,17681.0,36050.0,45559.0,708,3,2.58,True
1937,1354,34,59,1985,4,Male,7401 Seventh Street,35.46,-86.08,16349.0,33336.0,48213.0,692,3,2.95,True


In [36]:
user_df[user_df['retirement_age'] < 60]['gender'].value_counts()

gender
Female    49
Male      42
Name: count, dtype: int64

In [37]:
user_df[user_df['retirement_age'] >= 60]

,id,current_age,retirement_age,birth_year,birth_month,gender,address,latitude,longitude,per_capita_income,yearly_income,total_debt,credit_score,num_credit_cards,Debt_income_ratio,high_risk
0,825,53,66,1966,11,Female,462 Rose Lane,34.15,-117.76,29278.0,59696.0,127613.0,787,5,4.36,True
1,1746,53,68,1966,12,Female,3606 Federal Boulevard,40.76,-73.74,37891.0,77254.0,191349.0,701,5,5.05,True
2,1718,81,67,1938,11,Female,766 Third Drive,34.02,-117.89,22681.0,33483.0,196.0,698,5,0.01,False
3,708,63,63,1957,1,Female,3 Madison Street,40.71,-73.99,163145.0,249925.0,202328.0,722,4,1.24,True
4,1164,43,70,1976,9,Male,9620 Valley Stream Drive,37.76,-122.44,53797.0,109687.0,183855.0,675,1,3.42,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1995,986,32,70,1987,7,Male,6577 Lexington Lane,40.65,-73.58,23550.0,48010.0,87837.0,703,3,3.73,True
1996,1944,62,65,1957,11,Female,2 Elm Drive,38.95,-84.54,24218.0,49378.0,104480.0,740,4,4.31,True
1997,185,47,67,1973,1,Female,276 Fifth Boulevard,40.66,-74.19,15175.0,30942.0,71066.0,779,3,4.68,True
1998,1007,66,60,1954,2,Male,259 Valley Boulevard,40.24,-76.92,25336.0,54654.0,27241.0,618,1,1.08,True


In [38]:
user_df[user_df['retirement_age'] >= 60]['gender'].value_counts()

gender
Female    967
Male      942
Name: count, dtype: int64

In [39]:
def category_retirement_age(retirement_age):
    if 50 <= retirement_age <= 59:
        return 'Early'
    elif 60 <= retirement_age <= 65:
        return 'Normal'
    else:
        return 'Late'

user_df['retirement_Category'] = user_df['retirement_age'].apply(category_retirement_age)

In [40]:
user_df

,id,current_age,retirement_age,birth_year,birth_month,gender,address,latitude,longitude,per_capita_income,yearly_income,total_debt,credit_score,num_credit_cards,Debt_income_ratio,high_risk,retirement_Category
0,825,53,66,1966,11,Female,462 Rose Lane,34.15,-117.76,29278.0,59696.0,127613.0,787,5,4.36,True,Late
1,1746,53,68,1966,12,Female,3606 Federal Boulevard,40.76,-73.74,37891.0,77254.0,191349.0,701,5,5.05,True,Late
2,1718,81,67,1938,11,Female,766 Third Drive,34.02,-117.89,22681.0,33483.0,196.0,698,5,0.01,False,Late
3,708,63,63,1957,1,Female,3 Madison Street,40.71,-73.99,163145.0,249925.0,202328.0,722,4,1.24,True,Normal
4,1164,43,70,1976,9,Male,9620 Valley Stream Drive,37.76,-122.44,53797.0,109687.0,183855.0,675,1,3.42,True,Late
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1995,986,32,70,1987,7,Male,6577 Lexington Lane,40.65,-73.58,23550.0,48010.0,87837.0,703,3,3.73,True,Late
1996,1944,62,65,1957,11,Female,2 Elm Drive,38.95,-84.54,24218.0,49378.0,104480.0,740,4,4.31,True,Normal
1997,185,47,67,1973,1,Female,276 Fifth Boulevard,40.66,-74.19,15175.0,30942.0,71066.0,779,3,4.68,True,Late
1998,1007,66,60,1954,2,Male,259 Valley Boulevard,40.24,-76.92,25336.0,54654.0,27241.0,618,1,1.08,True,Normal


In [41]:
user_df['retirement_Category'].value_counts()

retirement_Category
Late      1234
Normal     675
Early       91
Name: count, dtype: int64

In [42]:
# cards Dataset
cards_df.head()

,id,client_id,card_brand,card_type,card_number,expires,cvv,has_chip,num_cards_issued,credit_limit,acct_open_date,year_pin_last_changed,card_on_dark_web
0,4524,825,Visa,Debit,4344676511950444,12/2022,623,YES,2,$24295,09/2002,2008,No
1,2731,825,Visa,Debit,4956965974959986,12/2020,393,YES,2,$21968,04/2014,2014,No
2,3701,825,Visa,Debit,4582313478255491,02/2024,719,YES,2,$46414,07/2003,2004,No
3,42,825,Visa,Credit,4879494103069057,08/2024,693,NO,1,$12400,01/2003,2012,No
4,4659,825,Mastercard,Debit (Prepaid),5722874738736011,03/2009,75,YES,1,$28,09/2008,2009,No


In [43]:
cards_df['client_id'].nunique()

2000

In [44]:
user_df['id'].nunique()

2000

In [50]:
user_df.rename(columns = {'id' :'client_id'}, inplace = True)
user_df.head(1)

,client_id,current_age,retirement_age,birth_year,birth_month,gender,address,latitude,longitude,per_capita_income,yearly_income,total_debt,credit_score,num_credit_cards,Debt_income_ratio,high_risk,retirement_Category
0,825,53,66,1966,11,Female,462 Rose Lane,34.15,-117.76,29278.0,59696.0,127613.0,787,5,4.36,True,Late


In [51]:
cards_dff = cards_df.groupby('client_id').agg({'card_brand' : 'count' , 'has_chip': 'max'}).reset_index()
cards_dff

,client_id,card_brand,has_chip
0,0,4,YES
1,1,3,YES
2,2,5,YES
3,3,4,YES
4,4,5,YES
...,...,...,...
1995,1995,4,YES
1996,1996,3,YES
1997,1997,7,YES
1998,1998,3,YES


In [52]:
cards_dff['has_chip'].value_counts()

has_chip
YES    1958
NO       42
Name: count, dtype: int64

In [53]:
total_user_card = pd.merge(user_df, cards_dff, on = 'client_id', how = 'left')
total_user_card

,client_id,current_age,retirement_age,birth_year,birth_month,gender,address,latitude,longitude,per_capita_income,yearly_income,total_debt,credit_score,num_credit_cards,Debt_income_ratio,high_risk,retirement_Category,card_brand,has_chip
0,825,53,66,1966,11,Female,462 Rose Lane,34.15,-117.76,29278.0,59696.0,127613.0,787,5,4.36,True,Late,5,YES
1,1746,53,68,1966,12,Female,3606 Federal Boulevard,40.76,-73.74,37891.0,77254.0,191349.0,701,5,5.05,True,Late,5,YES
2,1718,81,67,1938,11,Female,766 Third Drive,34.02,-117.89,22681.0,33483.0,196.0,698,5,0.01,False,Late,5,YES
3,708,63,63,1957,1,Female,3 Madison Street,40.71,-73.99,163145.0,249925.0,202328.0,722,4,1.24,True,Normal,4,YES
4,1164,43,70,1976,9,Male,9620 Valley Stream Drive,37.76,-122.44,53797.0,109687.0,183855.0,675,1,3.42,True,Late,1,YES
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1995,986,32,70,1987,7,Male,6577 Lexington Lane,40.65,-73.58,23550.0,48010.0,87837.0,703,3,3.73,True,Late,3,YES
1996,1944,62,65,1957,11,Female,2 Elm Drive,38.95,-84.54,24218.0,49378.0,104480.0,740,4,4.31,True,Normal,4,YES
1997,185,47,67,1973,1,Female,276 Fifth Boulevard,40.66,-74.19,15175.0,30942.0,71066.0,779,3,4.68,True,Late,3,YES
1998,1007,66,60,1954,2,Male,259 Valley Boulevard,40.24,-76.92,25336.0,54654.0,27241.0,618,1,1.08,True,Normal,1,NO


In [54]:
total_user_card[['per_capita_income', 'card_brand']]

,per_capita_income,card_brand
0,29278.0,5
1,37891.0,5
2,22681.0,5
3,163145.0,4
4,53797.0,1
...,...,...
1995,23550.0,3
1996,24218.0,4
1997,15175.0,3
1998,25336.0,1


In [55]:
total_user_card[['credit_score','card_brand']].sort_values(by = 'credit_score', ascending = False)

,credit_score,card_brand
490,850,6
1729,850,1
1313,850,6
773,850,4
1527,850,5
...,...,...
806,491,3
1510,490,2
908,489,3
150,488,2


In [56]:
total_user_card.to_csv("Total_User_Cards.csv", index = False)

Data Summary

First user_df:
. There are around 2000 users in the dataset, while the cards dataset contains about 6000 cards. This means that most users have multiple cards
. retirement_ages are mostly between 60–65 years with no users below 50
. A large portion of users have a debt-to-income ratio > 1. This means that their debt exceeds their income
. Credit_Score ranges from 480 to 850 with a significant group achieving the maximum score of 850

Second cards_df:
. Most users have multiple cards
. There is Variety in card brands (Visa, Mastercard,..)
. Most cards have chips, but some users still use non-chip cards

Third Total_user_card:
. Most users with higher incomes have more cards
. Users with high credit_score (close to 850) have more or better cards
